### ISSUE TITLE
Use type context from synthesized `TypedDict` constructor when inferring arguments passed to that constructor

### ISSUE THREAD

**Issue by niksabaldun** [astral-sh/ty#3027]
### Summary

```py
from __future__ import annotations

from typing import Any, NotRequired, TypeAlias, TypedDict


class Comparison(TypedDict):
    field: str
    op: NotRequired[str]
    value: Any

class Logical(TypedDict):
    op: NotRequired[str]
    conditions: list[Filter]

Filter: TypeAlias = Comparison | Logical

logical = Logical(conditions=[Comparison(field='a', value='b')])

```

On the last line, ty complains: Invalid argument to key "conditions" with declared type `list[Comparison | Logical]` on TypedDict `Logical`: value of type `list[Comparison]`. However, the list was never declared as `list[Comparison]`; ty infers it because it currently contains only Comparison objects. But it is perfectly compatible with `list[Comparison | Logical]`, so there should be no error reported.

### Version

0.0.21

**Comment by AlexWaygood** [comment]
Thanks for the report! Here's a slightly more minimal repro:

```py
from __future__ import annotations

from typing import TypedDict

class Comparison(TypedDict):
    field: str

class Logical(TypedDict):
    conditions: list[Comparison | Logical]

logical = Logical(conditions=[Comparison(field='a')])
```

it looks like this bug is specific to the (synthesized) constructor methods of `TypedDict` classes -- we emit no errors on this similar snippet:

```py
from __future__ import annotations

from typing import TypedDict

class Comparison(TypedDict):
    field: str

class Logical(TypedDict):
    conditions: list[Comparison | Logical]

class Foo:
    def __init__(self, conditions: list[Comparison | Logical]): ...

foo = Foo(conditions=[Comparison(field='a')])
```
